### PR TITLE
Another BuildPackages.sh overhaul

### PR DESCRIPTION
**Pull request by fingolfin**
This is another alternative to PRs #1121, #1122, #1123, #1127. It addresses issues #1120 and #1126.

### CODE REVIEW

File: bin/BuildPackages.sh [hunk 1]
```diff
@@ -60,36 +30,51 @@ then
...
-      shift
-    ;;
-
-    *)
-      cd ..
-      GAPDIR="$(pwd)"
-      cd "$CURDIR"
-    ;;
-  esac
+  PACKAGES="$(find . -maxdepth 2 -type f -name PackageInfo.g  | xargs -n 1 dirname)"
```
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: Why is it better to parse the output of ````find```` than to use 
````
shopt -s nullglob
PACKAGES=(*)
shopt -u nullglob
````
which should be the proper way to get all directories. Or you do not want to do the ````PackageInfo.g```` check later?

Comment (fingolfin) [inline bin/BuildPackages.sh]: The code using nullglob does not give all directories, it gives *all files*. But we only want directories that contain `PackageInfo.g`. 

And I still want to do the later check -- but now I can use it to valide that user arguments: I.e. if the user accidentally specified an invalid package directory name, we can show a warning (I actually forgot to add this, but added it now -- thanks for the poke).
Note that I really want this, because it actually bit me before (e.g. I entered `$GAPROOT/bin/BuildPackages.sh io orb` and wondered that nothing happened)

**Comment by hungaborhorvath** [comment]
@fingolfin I did more or less the same in #1123 in basically the same time you did this. Too bad.... :-( 

I guess it makes more sense to merge your PR rather than mine, but maybe you may find some things there that can be useful if you take a look? Definitely makes no sense to review all PRs about this no-coloring issue.

**Comment by fingolfin** [comment]
@hungaborhorvath Yes, sorry about the duplicate work -- as you said, we worked on it simultaneously, I wouldn't have started it if I had realized you worked on it already (I guess I should have *expected* it, but I honestly did not consider it).

I have to do some other work now, we can discuss how to proceed later -- e.g. as you suggest, I could incorporate some of your improvements here, like the --no-colour alternative spellings for british people :-).

**Comment by hungaborhorvath** [comment]
Maybe collecting some parts into functions, and the main body into another function to preserve the logging into one file. Sometimes it can be easier to check only one file, but of course ultimately everything important will be in the particular $PKG.log files. 

Also, yes, the --no-color alternatives, and maybe the logdir and logfile atlternatives.

File: bin/BuildPackages.sh [hunk 2]
```diff
@@ -130,8 +109,12 @@ notice \
 "Note that many GAP packages require extra programs to be installed," \
 "and some are quite difficult to build. Please read the documentation for" \
 "packages which fail to build correctly, and only worry about packages" \
-"you require!" \
-"Logging into ./$LOGDIR/$LOGFILE.log"
+"you require!"
+
+echo_run() {
+  echo "$@"
```
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: Is this for testing pruposes? Will it be later removed from the script? If not, then maybe we could add something to make clear that this is now a command we are running instead of an output of some command. E.g.
````echo "Running     " "$@"````

Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: Or even better, maybe use ````notice```` here.

File: bin/BuildPackages.sh [hunk 3]
```diff
@@ -236,17 +209,13 @@ build_one_package() {
       build_carat
     ;;
 
-    cohomolo*)
```
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: Is the ````cohomolo```` case not needed anymore? I have no idea, not using it, just (double) checking....

File: bin/BuildPackages.sh [hunk 4]
```diff
@@ -60,36 +30,52 @@ then
...
+[[ -t 1 ]] || COLORS=no
+
+while [[ "$#" -ge 1 ]]; do
+  option="$1" ; shift
+  case "$option" in
+    --with-gaproot)   GAPDIR="$1"; shift ;;
+    --with-gaproot=*) GAPDIR=${option#--with-gaproot=}; shift ;;
+    --no-color)       COLORS=no ;;
+    -*)               echo "ERROR: unsupported argument $option" ; exit 1;;
+    *)                PACKAGES="$PACKAGES $option" ;;
```
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: I prefer to use ````PACKAGES+=("$1")````, because I burnt myself with this one once. I do not recall exactly how, though....

File: bin/BuildPackages.sh [hunk 5]
```diff
@@ -279,8 +248,8 @@ build_one_package() {
   ) || build_fail
 }
 
-date >> "$LOGDIR/$FAILPKGFILE.log"
-for PKG in "${PACKAGES[@]}"
+date >> "$LOGDIR/fail.log"
+for PKG in ${PACKAGES}
```
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: Hm, I did not know this also works.

Comment (ChrisJefferson) [inline bin/BuildPackages.sh]: The advantage of `PACKAGES+=("$1")` and `for PKG in "${PACKAGES[@]}"` is that it will support directory names with spaces in.

Comment (fingolfin) [inline bin/BuildPackages.sh]: I am fully aware of this, but package names are not allowed to contain spaces, and if any package dir ever contains a space, lots of other things will break. So all in all, I favor this well-known idiom over using bash arrays, which are in comparision an obscure feature.

File: bin/BuildPackages.sh (hunk 3 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: Not only is it not needed, the special code failed to work because there is no `makefile.orig` in `cohomolo`.

File: bin/BuildPackages.sh (hunk 2 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: This does exactly what it name suggests: It echos a command before running it. This is a limited emulation of what set -x / -v do; it is used to output the exact configure command that is being run, which is useful for the log and for debugging build issues. It allows you to see which flags were passed to configure exactly. It is meant to stay.

Sure, this could use `notice` or add the prefix "Running " but personally I wouldn't find it  useful. A matter of taste, I guess

File: bin/BuildPackages.sh (hunk 5 above)
Comment (ChrisJefferson) [inline bin/BuildPackages.sh]: I'm happy to forbid this, but on a brief experiment (I just renamed all my package directories to have a space in), everything seems to work fine. We could at some point add an explicit check to reject such packages from `LoadPackage` (or somewhere similar).

File: bin/BuildPackages.sh (hunk 3 above)
Comment (ChrisJefferson) [inline bin/BuildPackages.sh]: That code was kept from the old `BuildPackages.sh` unchanged. I'm happy to believe it was bad, I should have checked when I ported it across. i don't use the package.

File: bin/BuildPackages.sh (hunk 4 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: I prefer writing non-bash shell code and avoiding bash extensions like arrays were possible... Though they are useful when one wants to deal with spaces in filenames... which we don't do here.

Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: The pkg directory (for other reasons) may contain other directories with spaces, even though they should not. 

Anyway, could you at least make a comment here that package directories are not supposed to have spaces or other special characters in their names?

File: bin/BuildPackages.sh (hunk 5 above)
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: Again, at least a comment would be nice here about packages should not have spaces, so it is safe to do this.

File: bin/BuildPackages.sh [hunk 6]
```diff
@@ -112,7 +112,7 @@ notice \
 "you require!"
 
 echo_run() {
-  echo "$@"
+  notice "Running $(printf "'%s' " "$@")"
```
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: Why is ````notice "Running" "$@"```` not enough here? Is it because of some escaping issue? 

I am also wondering if this code runs correctly as it is. I would think that there should only be a " before ````Running```` and one after the closing parenthesis (two altogether in the whole line, but definitely not four cut up by the parentheses).

File: bin/BuildPackages.sh [hunk 7]
```diff
@@ -60,36 +30,52 @@ then
...
 fi
 
-# CURDIR
 CURDIR="$(pwd)"
+GAPDIR="$(cd .. && pwd)"
+COLORS=yes
+PACKAGES=
+
+# If output does not go into a terminal (but rather into a log file),
+# turn of colors.
```
Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: of -> off

Comment (hungaborhorvath) [inline bin/BuildPackages.sh]: I was also wondering about the style of the comments. It would be nice to unify them. Currently some comments are complete sentences, some are only words, some are half sentences starting with small letters, etc. This of course need not be part of this PR.

File: bin/BuildPackages.sh (hunk 6 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: No, that simply notice line would not be enough. Note the printf is called with a format string that only caputes a single argument. In that situation, printf applies the format string to each argument in sequence. It thus prints out each argument quoted with single quotes. That means you can tell the difference between  `./configure CFLAGS=-m32 LDFLAGS=-m32` (which gets printed as  `'./configure' 'CFLAGS=-m32' 'LDFLAGS=-m32'` and the incorrect ` `./configure "CFLAGS=-m32 LDFLAGS=-m32"` (which gets printed as  `'./configure' 'CFLAGS=-m32 LDFLAGS=-m33'`)

As to why the code runs correct: Because those are the (admitedly bizarre) rules of POSIX shell quoting...

File: bin/BuildPackages.sh (hunk 7 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: In my eyes, comments need to be useful, that's the primary measure. I don't believe in unifying for the sake of unifying. If there is a concrete comment you think that is not helpful (because its too short, too long, misses the point, etc.), please point it out.

File: bin/BuildPackages.sh (hunk 3 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: I am not really using it myself -- I just run `BuildPackages.sh` and looked into all the errors it produced :-).

File: bin/BuildPackages.sh (hunk 5 above)
Comment (fingolfin) [inline bin/BuildPackages.sh]: Most unix shell scripts are horribly broken with regards to whitespace. E.g. the current BuildPackages.sh script is, too, because it incorrectly quotes in various places.

Anyway, I'll add back the arrays to make you guys happy, but again: this won't make it work with spaces in filenames, because other places in this script are doing it wrong (just try it)

Comment (fingolfin) [inline bin/BuildPackages.sh]: Plus, package build systems are quite likely to break due to spaces in their absolute paths, too...
And I think no regular user will try to rename their package directories, let alone insert spaces... so this feels a lot like hypothetical bike shedding to me... Anyway, as I said, I'll restore the array code, then if anybody feels this is something important, they can add whitespace-paths upport in another PR (but please actually test it)

Comment (ChrisJefferson) [inline bin/BuildPackages.sh]: We should at some point make an "official" decision on if we support packages with spaces in or not -- I don't mind banning them, but I get annoyed by software where spaces in filenames kind-of works in some places, then crashes horribly in others. Once we've made a decision for GAP as a whole, we can fix the script one way or the other -- but that pull request can come later.

Comment (fingolfin) [inline bin/BuildPackages.sh]: When you say "packages with spaces", I assume you mean "package directories with spaces", and then I think you should drop the "package" and just ask for support for directories with spaces... I suggest opening an issue for that.

Comment (fingolfin) [inline bin/BuildPackages.sh]: BTW, I think we currently do not restrict package *names* in any way (of course the package name does not have to be related to the package directory name at all, it's just a useful convention). We might want to revise that, too, and e.g. insist that it is a valid identifier (as in `IsValidIdentifier`). I kind of thought that's necessary for our namespaces anyway, but `ValidatePackageInfo` doesn't check it, it seems? @alex-konovalov ?

**Comment by fingolfin** [comment]
Seems the new changes I pushed made some of the old code comments here vanish... so I'll try to summarize their content here:

1. Regarding spaces "in packages": I assume this was meant to refer to "spaces in package directory names", and I suggested that there is no reason to stop there, and @ChrisJefferson might want to submit an issue in general about support for spaces in directory names -- not just in this script, but also in the build system (I have no idea if it works with that), in other scripts (gac? docomp? ...) and possibly also in GAP itself (though I am hopeful that it is doing fine in that regard)

2. This script, even when having `PACKAGES` a bash array, is not working correctly when a package directory name contains spaces, e.g. the log code does not quote `$PKG` correctly in all places.

3. To my surprise I (re)discovered that package names seem to have no restriction in GAP, at least `ValidatePackageInfo` does not require anything. Perhaps we should require `IsValidIdentifier` for them? I wonder what happens to the namespace feature if a package has a space in its name?

**Comment by ChrisJefferson** [comment]
This patch looks good to me, are you ( @fingolfin , @hungaborhorvath ) happy for it to be merged?

**Comment by alex-konovalov** [comment]
@fingolfin at the moment package name is required to be a non-empty string. That was probably used before `IsValidIdentifier` appeared (although I may be wrong here). The Example package manual says [here](https://gap-packages.github.io/example/doc/chapA.html#X8383876782480702) that 

> A GAP package should have an alphanumeric name (package-name, say); mixed case is fine, but there should be no whitespaces. All files of a GAP package package-name must be collected in a single directory package-dir, where package-dir should be just package-name preferably converted to lowercase and optionally followed by the package version (with or without hyphen to separate the version from package-name).

It's useful to update `ValidatePackageInfo` to actually check that.

**Comment by hungaborhorvath** [comment]
@ChrisJefferson @fingolfin Yes, I do not mind if it is merged, at all. Could you please close the other relevant PRs, then? Thanks.

**Comment by alex-konovalov** [comment]
Analysing which package are not loadable after merging this. `make testpackagesload` reports float, fplsa, help, itc, kbmag, linboxing, normalizinterface, pargap, polymakeinterface and xgap. Clearly, xgap, pargap and itc are not of a concern, polymake is not yet installed on that slave, and linboxing does not work anywhere. Genuine failures:

# float
```
checking for fplll.h... no
checking for lllReduction in -lfplll... no
./configure: line 19524: test: too many arguments
configure: error: Compilation of the external libraries  mpfr mpfi mpc cxsc fplll was requested, but is now disabled. Please install them manually and rerun configure.
Running 'gmake' 
gmake: *** No targets specified and no makefile found.  Stop.

WARNING: Failed to build float-0.7.4
```

# fplsa
```
==== Checking fplsa
Running './configure' '/data/gap-jenkins/workspace/GAP-pkg-update-master-quicktest/GAPCOPTS/64build/GAPGMP/gmp/GAPTARGET/packagesload/label/kovacs/GAP-pkg-update-master-snapshot' 
Running 'gmake' 'CC=gcc -O2 ' 
gmake: invalid option -- 'O'
gmake: invalid option -- '2'
Usage: gmake [options] [target] ...
```

# kbmag
```
==== Checking kbmag-1.5.4
Running './configure' '/data/gap-jenkins/workspace/GAP-pkg-update-master-quicktest/GAPCOPTS/64build/GAPGMP/gmp/GAPTARGET/packagesload/label/kovacs/GAP-pkg-update-master-snapshot' 
Running 'gmake' 'COPTS=-O2 -g' 
gmake: invalid option -- 'g'
```

# NormalizInterface
```
==== Checking NormalizInterface-0.9.8
Found GAP in directory /data/gap-jenkins/workspace/GAP-pkg-update-master-quicktest/GAPCOPTS/64build/GAPGMP/gmp/GAPTARGET/packagesload/label/kovacs/GAP-pkg-update-master-snapshot
GAParch = x86_64-pc-linux-gnu-gcc-default64
GAP was built with its own GMP
GAP's GMP includes C++ support
Fetching Normaliz source code
Cloning into 'Normaliz.git'...
Note: checking out 'fdf877804e473a4f59e3b20465598f97088fb547'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b new_branch_name

./build-normaliz.sh: line 94: cmake: command not found
```
